UpdatedImageProcessing/Classifiers/classify_color.py

```diff
@@ -0,0 +1,220 @@
+		}
+
+	def get_color(self, img=None):
+		"""
+		Gets the color of the background and text inside a target.
+
+		If no img is passed to get_color, then this function assumes to use the 
+		PIL image passed in the constructor.
+
+		:param img:		optional. Opened PIL image.
+		:type img:		PIL object.
+
+		:return:		Two element list, where the first element is the shape
+						color, and the second element is the text color. Ex:
+							["red", "green"]
+						or, simply:
+							[<background color>, <text color>]
+		"""
+		passed = Image.open(self.img)
+
+		# convert pil to opencv2
+		img = np.array(passed)
+		img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+
+		Z = img.reshape((-1, 3))
+
+		# convert to np.float32
+		Z = np.float32(Z)
+
+		# define criteria, number of clusters(K) and apply kmeans()
+		criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 10, 1.0)
+		K = 3
+		ret, label, center = cv2.kmeans(Z, K, None, criteria, 10, cv2.KMEANS_RANDOM_CENTERS)
+
+		# now convert back into uint8, and make original image
+		center = np.uint8(center)
+		res = center[label.flatten()]
+		res2 = res.reshape((img.shape)) # actual image
+
+		# grab colors from image
+		res3 = cv2.cvtColor(res2, cv2.COLOR_RGB2BGR)
+		colors_from_img = set( tuple(v) for m2d in res3 for v in m2d )
+
+		print("classify_color.py: Raw colors from image: " + str(colors_from_img))
+
+		color_names_from_img = []
+
+		for color in colors_from_img:
+			color_names_from_img.append(self.convert_rgba_to_color_name(list(color)))
+
+		if "white" in color_names_from_img:
+			color_names_from_img.remove("white") # remove background
+
+		print("classify_color.py: Converted color names: " + str(color_names_from_img))
+
+		# figure out which color comes first, reading from the center row outwards
+		height, width, channels = img.shape
+		print("classify_color.py: Height of image is " + str(height) + "px and width is " + str(width) + "px.")
+		print("classify_color.py: Height divided by 2, rounded, is " + str(int(round(height/2.0))) + ".")
+
+		row = int(round(height/2.0))
+
+		shape_color = None
+		text_color = None
+
+		for column in range(1, width):
+			if self.convert_color_name_to_rgba(self.convert_rgba_to_color_name(res3[row, column])) == self.convert_color_name_to_rgba(color_names_from_img[0]):
+				shape_color = color_names_from_img[0]
+				text_color = color_names_from_img[1]
+			elif self.convert_color_name_to_rgba(self.convert_rgba_to_color_name(res3[row, column])) == self.convert_color_name_to_rgba(color_names_from_img[1]):
+				shape_color = color_names_from_img[1]
+				text_color = color_names_from_img[0]
+
+			if shape_color != None and text_color != None:
+				print("classify_color.py: Determined shape color to be " + str(shape_color) + ".")
+				print("classify_color.py: Determined text color to be " + str(text_color) + ".")
+				break;
+
+		result = [shape_color, text_color]
+
+		return result
+
+	def three_dimesional_distance(self, p1, p2):
```

Comment:
You can do this in an n-dimensional space easily with numpy. See the SDA package's ```VectorMath``` class for how to do it.

Reply:
[We looked into this, but the numpy solution I found online doesn't address the problem.](https://stackoverflow.com/a/1401828)

I'll look into [VectorMath.](https://github.com/FlintHill/SUAS-Competition/blob/master/SDAPackage/SDA/vector_math.py)

Reply:
Yes, this will do exactly what you need.